# Overview
- Train an Autogluon predictor (autoML)


In [22]:
# Import necessary packages
from utils import pd, np, TabularPredictor, accuracy_score

In [23]:
# Import data 
train_df = pd.read_csv("../data/processed/train_data.csv")
test_df = pd.read_csv("../data/processed/test_data.csv")

In [24]:
# Set random seed for reproducibility
np.random.seed(42)

In [25]:
# Initialize autogluon tabular predictor
predictor = TabularPredictor(
    label='label',
    eval_metric='accuracy',
    path='../models/autogluon'
)

In [26]:
# Train autogluon tabular predictor
predictor.fit(
    train_data=train_df,
    time_limit=600,  # 10 minutes time limit
    presets='medium_quality_faster_train',  # Use medium quality preset for faster training
    num_bag_folds=5,  # Number of folds for bagging
    num_stack_levels=2,  # Number of stacking levels
    verbosity=2,  # Show detailed training progress
)

Preset alias specified: 'medium_quality_faster_train' maps to 'medium_quality'.
Verbosity: 2 (Standard Logging)
=================== System Info ===================
AutoGluon Version:  1.3.1
Python Version:     3.9.6
Operating System:   Darwin
Platform Machine:   arm64
Platform Version:   Darwin Kernel Version 24.5.0: Tue Apr 22 19:54:49 PDT 2025; root:xnu-11417.121.6~2/RELEASE_ARM64_T6000
CPU Count:          10
Memory Avail:       34.22 GB / 64.00 GB (53.5%)
Disk Space Avail:   853.49 GB / 926.35 GB (92.1%)
Presets specified: ['medium_quality_faster_train']
Beginning AutoGluon training ... Time limit = 600s
AutoGluon will save models to "/Users/miriamlandau/Documents/predict_hand_imagery/models/autogluon"
Train Data Rows:    168
Train Data Columns: 258
Label Column:       label
AutoGluon infers your prediction problem is: 'binary' (because only two unique label-values observed).
	2 unique label values:  [np.int64(0), np.int64(1)]
	If 'binary' is not the correct problem_type, please man

In [30]:
# Evaluate model performance
performance = predictor.evaluate(test_df)
performance

{'accuracy': 0.7619047619047619,
 'balanced_accuracy': np.float64(0.7716346153846154),
 'mcc': np.float64(0.5282450493547782),
 'roc_auc': np.float64(0.78125),
 'f1': 0.7222222222222222,
 'precision': 0.65,
 'recall': 0.8125}

In [31]:
# Display autogluon leadorboard
predictor.leaderboard(test_df)

model  score_test  score_val eval_metric  \
0    NeuralNetFastAI_BAG_L3    0.785714   0.797619    accuracy   
1     NeuralNetTorch_BAG_L1    0.761905   0.779762    accuracy   
2           CatBoost_BAG_L2    0.761905   0.821429    accuracy   
3            XGBoost_BAG_L2    0.761905   0.797619    accuracy   
4   RandomForestGini_BAG_L2    0.761905   0.732143    accuracy   
5    NeuralNetFastAI_BAG_L2    0.761905   0.809524    accuracy   
6           LightGBM_BAG_L3    0.761905   0.839286    accuracy   
7       WeightedEnsemble_L4    0.761905   0.839286    accuracy   
8           CatBoost_BAG_L3    0.761905   0.821429    accuracy   
9   RandomForestGini_BAG_L3    0.761905   0.755952    accuracy   
10  RandomForestEntr_BAG_L3    0.761905   0.761905    accuracy   
11        LightGBMXT_BAG_L1    0.738095   0.785714    accuracy   
12   NeuralNetFastAI_BAG_L1    0.738095   0.803571    accuracy   
13      WeightedEnsemble_L2    0.738095   0.809524    accuracy   
14          LightGBM_BAG_L2    0.738095   0.821429    accuracy   
15     LightGBMLarge_BAG_L2    0.738095   0.779762    accuracy   
16        LightGBMXT_BAG_L3    0.738095   0.809524    accuracy   
17     LightGBMLarge_BAG_L3    0.738095   0.773810    accuracy   
18           XGBoost_BAG_L3    0.738095   0.827381    accuracy   
19    ExtraTreesGini_BAG_L3    0.738095   0.755952    accuracy   
20    ExtraTreesEntr_BAG_L3    0.738095   0.738095    accuracy   
21    NeuralNetTorch_BAG_L3    0.738095   0.767857    accuracy   
22          LightGBM_BAG_L1    0.714286   0.767857    accuracy   
23          CatBoost_BAG_L1    0.714286   0.797619    accuracy   
24           XGBoost_BAG_L1    0.714286   0.761905    accuracy   
25        LightGBMXT_BAG_L2    0.714286   0.827381    accuracy   
26      WeightedEnsemble_L3    0.714286   0.827381    accuracy   
27    ExtraTreesEntr_BAG_L2    0.714286   0.767857    accuracy   
28  RandomForestEntr_BAG_L2    0.714286   0.750000    accuracy   
29    NeuralNetTorch_BAG_L2    0.714286   0.773810    accuracy   
30  RandomForestEntr_BAG_L1    0.690476   0.684524    accuracy   
31    ExtraTreesGini_BAG_L2    0.690476   0.738095    accuracy   
32    ExtraTreesEntr_BAG_L1    0.619048   0.666667    accuracy   
33    ExtraTreesGini_BAG_L1    0.619048   0.714286    accuracy   
34  RandomForestGini_BAG_L1    0.619048   0.696429    accuracy   
35    KNeighborsUnif_BAG_L1    0.595238   0.440476    accuracy   
36     LightGBMLarge_BAG_L1    0.595238   0.767857    accuracy   
37    KNeighborsDist_BAG_L1    0.571429   0.505952    accuracy   

    pred_time_test  pred_time_val   fit_time  pred_time_test_marginal  \
0         0.584514       0.610228  29.397866                 0.037381   
1         0.110924       0.112273   2.304106                 0.110924   
2         0.283238       0.309759  19.550223                 0.015824   
3         0.294491       0.295251  12.816848                 0.027076   
4         0.299389       0.326428  12.416779                 0.031975   
5         0.315873       0.313011  14.535238                 0.048459   
6         0.556002       0.587927  27.292856                 0.008869   
7         0.556649       0.588331  27.330467                 0.000648   
8         0.562952       0.611860  32.951194                 0.015819   
9         0.577109       0.625844  27.236423                 0.029976   
10        0.579997       0.626904  27.164618                 0.032864   
11        0.010162       0.008447   0.478543                 0.010162   
12        0.038001       0.026358   2.434548                 0.038001   
13        0.070687       0.070567   2.844423                 0.000767   
14        0.276434       0.292140  12.675601                 0.009020   
15        0.280123       0.292311  13.760315                 0.012708   
16        0.555343       0.588062  27.278444                 0.008210   
17        0.560309       0.589524  28.453168                 0.013177   
18        0.572793       0.591717  27.484093                

In [32]:
# Display feature importance
feature_importance = predictor.feature_importance(train_df) 
feature_importance.sort_values(by='importance', ascending=False).head()
# runs too long

Computing feature importance via permutation shuffling for 258 features using 168 rows with 5 shuffle sets...
	898.08s	= Expected runtime (179.62s per shuffle set)


KeyboardInterrupt: 

# Key Takeways

- autogluon performs with an accuracy of 0.76 with a stacked ensemble
